system/security/shared/caching.hpp

```diff
@@ -151,16 +154,26 @@ class CPermissionsCache
 
     virtual ~CPermissionsCache();
 
-    static CPermissionsCache* queryInstance()
+    //Returns an owned reference to a shared cache of a given class type.
+    //Call this method with a unique class string ("LDAP", "MyOtherSecMgr")
+    //to create a cache shared amongst security managers of the same class
+    static CPermissionsCache* getInstance(const char * _secMgrClass)
     {
+        const char * secMgrClass = (_secMgrClass != nullptr  &&  *_secMgrClass) ? _secMgrClass : "genericSecMgrClass";
+
+        CriticalBlock block(PCCritSect);
+        MapCache::iterator it = g_mapCache.find(secMgrClass);
+        if (it != g_mapCache.end())//exists in cache
+        {
+            LINK((*it).second);
+            return (*it).second;
+        }
+        else
         {
-            CriticalBlock block(PCCritSect);
-            if (instance == nullptr)
-            {
-                instance = new CPermissionsCache();
-            }
+            CPermissionsCache * instance = new CPermissionsCache();
+            g_mapCache.insert(pair<string, CPermissionsCache*>(secMgrClass, instance));
+            return instance;
```

Comment:
Is the intention that the CPermissionCache objects only exist as long as anyone is using one of them? Or that they will persist indefinitely once created? Assuming the former, this code is probably ok but the CPermissionsCache destructor should (a) only remove its own entry from the map, and (b) should do so while in possession of the PCCritSect lock.